grid/crm/contacts/maps/crisp.suma

```diff
@@ -0,0 +1,286 @@
+
+  http PATCH "/v1/website/{parameters.WEBSITE_ID}/people/profile/{profileId}" {
+    security "plugin_tier"
+
+    request "application/json" {
+      headers { "X-Crisp-Tier" = "plugin" }
+      body = profileUpdate
+    }
+
+    response 200 {
+      set {
+        customPropsResult = call PatchCustomProperties(
+          profileId = profileId,
+          customProperties = input.customProperties
+        )
+      }
+
+      return map error if (!customPropsResult) {
+        title = "Partially failed"
+        detail = "Contact might be updated without custom properties. Please update the contact for data consistency"
+      }
+
+      map result {
+        id = profileId
+      }
+    }
+
+    response 409 {
+      map error {
+        title = "Email already exists"
+        detail = (`Email '${input.email}' already exists in another contact`)
+      }
+    }
+
+    response {
+      error = call HandleCommonErrors(statusCode = statusCode)
+      map error error
+    }
+  }
+}
+
+map Search {
+  OPERATOR_MAP = {
+    "EQ": 'eq',
+    "NEQ": 'neq'
+  }
+
+  http GET "/v1/website/{parameters.WEBSITE_ID}/people/profiles" {
+    security "plugin_tier"
+
+    request "application/json" {
+      query {
+        search_filter = [{
+          criterion : input.property,
+          query : [input.value],
+          operator : OPERATOR_MAP[input.operator] || input.operator,
+          model: "people"
+        }] 
+      }
+
+      headers { "X-Crisp-Tier" = "plugin" }
+    }
+
+    response 200 {
+      set {
+        contactsList = call foreach (profile of body.data) MapProfileFromCrisp(profile = profile)
+      }
+
+      map result contactsList
+    }
+
+    response {
+      error = call HandleCommonErrors(statusCode = statusCode)
+      map error error
+    }
+  }
+}
+
+operation GetCustomProperties {
+  http GET "/v1/website/{parameters.WEBSITE_ID}/people/data/{args.profileId}" {
+    security "plugin_tier"
+
+    request {
+      headers { "X-Crisp-Tier" = "plugin" }
+    }
+
+    response 200 {
+      return body.data.data
+    }
+
+    response {
+      error = call HandleCommonErrors(statusCode = statusCode)
+      fail error
+    }
+  }
+}
+
+operation PatchCustomProperties {
+  return if (Object.keys(args.customProperties).length < 1) { 
+    id = args.profileId
+  }
+
+  http PATCH "/v1/website/{parameters.WEBSITE_ID}/people/data/{args.profileId}" {
+    security "plugin_tier"
+
+    request {
+      headers { "X-Crisp-Tier" = "plugin" }
+      body = {
+        data: args.customProperties
+      }
+    }
+
+    response 200 { 
+      return { id = args.profileId }
+    }
+
+    response {
+      error = call HandleCommonErrors(statusCode = statusCode)
+      fail error
+    }
+  }
+}
+
+// Utilities
+operation BuildCrispProfile {
+  input = args.input
+
+  fullName = [input.firstName || "", input.lastName || ""].join(' ').trim()
+
+  return { 
+    email: input.email,
+
+    person: fullName.length || input.phone || input.country ? {
```

Comment:
This construction is pretty confusing 🤨

Reply:
Hmm, what exactly about it? 🤔  The fact that it's inlined, or the actual use of ternary for this?

The `person` object optional, but if it's present, some fields inside are required, that's why I need to always check whether to even construct it.

Alternatively I had bunch of `set if () {...}` for each field but it was one screen long, and not much more readable.

What do you have in mind for this?

Reply:
Well, the combination on inline and ternary is imo hard to read. Bunch of set if will be probably too long. 
It would be nice to do something like:
`set if(fullName.length || input.phone || input.country) {
  person = {...}
}`

Not sure that will work :/
Anyway it's not a blocker 🙂

Reply:
Those `set if() { person ... }` statements have problem that you cannot reference `person` if the condition evaluates to false. I need it to be undefined.

I was able to simplify it the following way, is it more readable?

```graphql
operation BuildCrispProfile {
  input = args.input

  fullName = [input.firstName || "", input.lastName || ""].join(' ').trim() || undefined

  return { 
    email: input.email,

    person: (fullName || input.phone || input.country) && {
      nickname: fullName,
      phone: input.phone,
      geolocation: { country: input.country }
    },

    company: input.company && {
      name: input.company
    }
  }
}
```

Reply:
Yep, I like that 👍👍